src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -67,6 +67,13 @@ internal partial class Http2Connection : IHttp2StreamLifetimeHandler, IHttpHeade
         private int _gracefulCloseInitiator;
         private int _isClosed;
 
+        private readonly Http2Stream[] _streamPool;
+        private int _pooledStreamCount;
+
+        // Since the number of streams per connection is user configurable, we need a max just in case that number is too big
+        // 200 seems reasonable since the default is 100
+        private static readonly int _maxPooledStreams = 200;
```

Comment:
I expect this limit can be a lot smaller in practice. Even with a high number of parallel requests you only need enough pool space to handle the turnover between requests completing and the next ones starting. Even a pool size of 5 or 10 could provide a significant benefit, and you wouldn't be holding on to a bunch of memory when the connection went idle. I'd be curious to see what the diminishing returns curve looks like.

Reply:
A pattern like what pipelines uses might be better here. https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegmentStack.cs

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L20-L21

Start the pool off with a minimum size and then pool up to a maximum size.